core: refactor `get_proc_name`

CuriousGeorgiy opened the pull request:
`get_proc_name` code is clogged with procedure cache management logic:
refactor this logic out.

Needed for #4002

NO_DOC:refactoring
NO_CHANGELOG:refactoring

locker commented on src/lib/core/backtrace.cc:
```diff
@@ -74,49 +74,75 @@ backtrace_proc_cache_clear(void)
+		return cache_name;
+
+	unw_get_proc_name(unw_cur, proc_name, sizeof(proc_name), offset);
+	backtrace_proc_cache_put(ip, proc_name, *offset);
```
You don't check the return value of `backtrace_proc_cache_put`. I'd drop the return value and panicked on memory allocation error (we do need `region_xalloc_object` similar to `xmalloc`...)

CuriousGeorgiy replied:
I looked at the `xalloc` wrappers you mention:
https://github.com/tarantool/tarantool/blob/18eae244c8e0fcd571c02438e4f9006b5bce4e67/src/trivia/util.h#L108-L121

Personally, considering we have C++ code, I despise calling `exit` in the middle of nowhere (i.e., panicking), quoting [cppreference](https://en.cppreference.com/w/cpp/utility/program/exit):
> Stack is not unwound: destructors of variables with automatic storage duration are not called.

Even if we had pure C code, local variables may require cleanup to be performed.

Anyways, I think the best solution we have at our hands now is `tnt_raise`, although at first glance it seems broken to me:
https://github.com/tarantool/tarantool/blob/18eae244c8e0fcd571c02438e4f9006b5bce4e67/src/lib/core/exception.h#L248-L253

Not only can `new` fail, when we are _out of memory_, but also the _catch site_ is burdened to deallocate this memory. AFAIC, we actually leak all memory allocated for _exceptions_: at least I didn't find any _catch site_ that would explicitly call `delete`. During my _GSOC_ project this summer, I once had to debug _Tarantool_ under _ASAN_ and it actually reported these leaks (not quite sure if it will currently, but I can try if needed).

Furthermore, there is a general guideline that _exceptions_ should be _thrown_ by value and _caught_ by reference: see [C++ Core Guidleines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-exception-ref) and an elaborate explanation in [C++ Coding Standards: 101 Rules, Guidelines, and Best Practices](http://library.bagrintsev.me/CPP/Sutter.C%2B%2B%20Coding%20Standards.2005.pdf).

Anyways, I would like to discuss this with you in person on Monday.

CuriousGeorgiy replied:
/cc @alyapunov @Korablev77

CuriousGeorgiy replied:
UPD: we do leak memory from exceptions


```Console
georgiy.lebedev@georgiy-lebedev src % ASAN_OPTIONS=detect_leaks=1 ./tarantool
Tarantool 2.10.0-beta2-90-g18eae244c
type 'help' for interactive help
tarantool> box.cfg{log='test:test'}
---
- error: 'Incorrect value for option ''log'': expecting a file name or a prefix, such
    as ''|'', ''pipe:'', ''syslog:'''
...

tarantool>
=================================================================
==54199==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 864 byte(s) in 1 object(s) allocated from:
    #0 0x11145d750 in wrap_malloc+0xa0 (libclang_rt.asan_osx_dynamic.dylib:x86_64+0x44750)
    #1 0x10da1edc4 in Exception::operator new(unsigned long) exception.cc:93
    #2 0x10d60716f in box_check_say() box.cc:701
    #3 0x10d606c3c in box_check_config() box.cc:1253
    #4 0x10d8ed350 in lbox_cfg_check(lua_State*) cfg.cc:50
    #5 0x10daaffba in lj_BC_FUNCC+0x43 (tarantool:x86_64+0x100819fba)
```

CuriousGeorgiy replied:
Fixed: came to an agreement that _panicking_ is legit under the circumstances it can actually happen. Anyways, `tnt_raise` is not a viable option here, as there is no C++ calling stack to catch this exception. Added `unlikely` annotation and `panic_syserror` in case of OOM.

locker reviewed (changes requested):
Please set yourself for the commit author:

```sh
git commit --amend --no-edit --reset-author
```

CuriousGeorgiy commented on src/lib/core/backtrace.cc:
```diff
@@ -74,49 +74,75 @@ backtrace_proc_cache_clear(void)
+	node.key = ip;
+	node.val = entry;
+	entry->offset = offset;
+	strncpy(entry->name, name, BACKTRACE_NAME_MAX);
```
Wow, I actually read the _cppreference_ page, but happened to miss this sentence:
> If count is reached before the entire array src was copied, the resulting character array is not null-terminated.

Shame on me, fixed.

CuriousGeorgiy commented on src/lib/core/backtrace.cc:
```diff
@@ -74,49 +74,75 @@ backtrace_proc_cache_clear(void)
+	strncpy(entry->name, name, BACKTRACE_NAME_MAX);
+
+	mh_int_t k = mh_i64ptr_put(proc_cache, &node, NULL, NULL);
+	if (k == mh_end(proc_cache)) {
```
Apparently, Egor wrote this code before your patch, I didn't pay attention to this snippet.

[_nit_] I guess the mention about mh_end should also be removed:
https://github.com/tarantool/tarantool/blob/18eae244c8e0fcd571c02438e4f9006b5bce4e67/src/lib/salad/mhash.h#L301-L302

Fixed.

CuriousGeorgiy commented on src/lib/core/backtrace.cc:
```diff
@@ -74,49 +74,75 @@ backtrace_proc_cache_clear(void)
-	struct proc_cache_entry *entry;
-	struct mh_i64ptr_node_t node;
-	mh_int_t k;
+	return -1;
```
I agree, this function's interface looks ugly, didn't pay attention to it: fixed.

CuriousGeorgiy commented on src/lib/core/backtrace.cc:
```diff
@@ -74,49 +74,75 @@ backtrace_proc_cache_clear(void)
+}
+
+const char *
+get_proc_name(unw_cursor_t *unw_cur, unw_word_t *offset, bool skip_cache)
```
I agree: fixed.

CuriousGeorgiy commented on src/lib/core/backtrace.cc:
```diff
@@ -74,49 +74,75 @@ backtrace_proc_cache_clear(void)
-const char *
-get_proc_name(unw_cursor_t *unw_cur, unw_word_t *offset, bool skip_cache)
+static int
+backtrace_proc_cache_find(unw_word_t ip, const char **name, unw_word_t *offset)
```
I would like to discuss this point.

From your point of view, could we provide _descriptive_ and _concise_ enough function names (if you consider the current ones are not), so that they would document themself and FWIW (DRY and documentation maintenance) save us from the need to provide such comments?

CuriousGeorgiy replied:
Came to an agreement that descriptive and concise comments are helpful and follows the _double-entry bookkeeping_ principle.

CuriousGeorgiy commented:
> Please set yourself for the commit author:
> 
> ```shell
> git commit --amend --no-edit --reset-author
> ```

Fixed: set myself as commit author and added _Co-Authored-By_ trailers to mark Egor's contribution.
